商品購入機能

NUKEDO opened the pull request:
# What
商品購入機能の実装
# Why
出品された商品を購入できるようにするため
取引を成立させるメイン機能のひとつであるため
## Github
- 「出品ユーザーが別であり、購入者のいない商品」のみ、購入ボタンが出現し購入ページへ移動できる
https://gyazo.com/00560cd50af46045dcd22c3194c35b6e
- ログインしていないユーザーは購入ボタンが表示されず、購入ページへも直接飛べない
https://gyazo.com/cc1a13e75219cf9c49239b97df3997b7
- ログインしているユーザーでも、購入済みの商品や自身で出品した商品の購入ページへは飛べない
https://gyazo.com/f89058d1399207f26fc4ec9f7f01ea42
- 購入画面で入力した情報が誤っている場合にはエラーが発生する・入力が正しければ購入できる
https://gyazo.com/676e05f9cd1cc9821689c6ead7177fab
- 単体テストコードの実行
https://gyazo.com/0f0d62c8904eac3f76e64a3251c9b839

NUKEDO committed:
customersコントローラーとモデルの作成

ルーティングも確認済

NUKEDO committed:
マイグレーション記述と実行

直し忘れてたREADMEのカラム名も修正
バリデーションはまだ

NUKEDO committed:
Update 20201102091700_create_customers.rb

NUKEDO committed:
Revert "Update 20201102091700_create_customers.rb"

This reverts commit 942d7184b1e0fb084241c79bc78fdf0fbd03cf37.

NUKEDO committed:
Revert "マイグレーション記述と実行"

This reverts commit 8fed46f08fa8e688597c6c282aad45cd48dc2b74.

NUKEDO committed:
Revert "customersコントローラーとモデルの作成"

This reverts commit 37f737e79f73aae7008fed8b76408f4831ff5d6c.

NUKEDO committed:
コントローラー名等をorderにして作成し直し

rails dとdb:rollback実行済み
rails g controller orders
ルーティング記述
ビューファイルとCSS配置
直し忘れてたREADME修正

NUKEDO committed:
indexをshowに変更

ルーティングとコントローラーのindexアクションをshowに変更
ビューとCSSのファイル名をshowに変更

NUKEDO committed:
purchaseテーブル作成、showビュー編集

purchaseテーブル作成
createアクション作成
showページが表示されるよう変更

NUKEDO committed:
ルーティングをネストの形に修正

合わせてshowアクションになっていたのをindexアクションに修正
purchase_recordにデータが作成されることを確認

NUKEDO committed:
purchase_record関連の部分で直してなかった部分を修正

NUKEDO committed:
customerモデル作成

アソシエーション記述
マイグレーション記述、db:migrate
ActiveHash設定
ビューを元に戻す

NUKEDO committed:
都道府県カラム名を修正

prefecture_idからprefectureに修正

NUKEDO committed:
購入機能実装(バリデーション 未完成)

NUKEDO committed:
各商品画像のSoldOut画像表示は商品状態で変化するよう修正

NUKEDO committed:
エラーメッセージが表示されるよう変更

NUKEDO committed:
バリデーション設定と単体テストコード作成

NUKEDO committed:
Rubocop実行

ky3090 commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,53 @@
+      @order_form.save
+      redirect_to root_path
+    else
+      @item = Item.find(params[:item_id])
```
5,23行目の```@item = Item.find(params[:item_id])```は同一ファイル内で記述が重複していますので、private以下にメソッドを定義し、before_actionで呼び出しましょう！

NUKEDO replied:
修正しました。

ky3090 commented on spec/factories/order_forms.rb:
```diff
@@ -0,0 +1,13 @@
+  factory :order_form do
+    token         { 'tok_abcdefghijk00000000000000000' }
+    postal_code   { '980-0001' }
+    prefecture_id { '5' }
```
prefecture_idは数値ですので、シングルクウォーテーションで囲まないようにしましょう！

NUKEDO replied:
修正しました。

ky3090 commented on spec/models/order_form_spec.rb:
```diff
@@ -0,0 +1,137 @@
+        expect(@order_form.errors.full_messages).to include('Prefecture Select')
+      end
+      it '都道府県が半角なので購入できない' do
+        @order_form.prefecture_id = '２２'
```
prefecture_idは数値ですので、シングルクウォーテーションで囲まないようにしましょう！

NUKEDO replied:
修正しました。

ky3090 reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

ky3090 commented:
```
- ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移すること
- ログイン・ログアウトの状態に関わらず、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移すること
```

キャプチャ画像ありがとうございます！
上記に関するキャプチャ画像もいただけますと幸いです。
よろしくお願いいたします！

NUKEDO committed:
コードレビューを受けて修正

コントロール内の同じ記述をprivateメソッドに変更
テストコード及びFactoryBot内のinteger指定カラムのクォーテーションを除外
Rubocop実行済、テストコード実行済
ブラウザによる正常購入確認

NUKEDO commented:
修正しました！
追加のキャプチャ画像です。ご確認ください。
- ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移すること
https://gyazo.com/37d5179da9c5a68499c3c0b9c9a3d053
- ログイン・ログアウトの状態に関わらず、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移すること
https://gyazo.com/afe2589d3c0c86a1ae6de7beb9fe79f3

iy06 commented on app/views/items/index.html.erb:
```diff
@@ -112,25 +112,30 @@
+              <% @stock = nil %>
+              <% break %>
+            <% end %>
+          <% end %>
```
122 - 138行目の記述は、商品が売れているかどうか判別する記述でしょうか？

もし、そうであれば
```
<% if item.purchase_record %> 
  <div class='sold-out'>
    <span>Sold Out!!</span>
  </div>
<% end %>
```
上記の記述で判断できそうです。

もし、ご不明な点がございましたらコネクトにてのご質問をお願いいたします。

NUKEDO replied:
どうするべきか悩んでいた箇所です、ありがとうございます！
itemsコントローラーとordersコントローラーにも同様に似た記述がありますので、
こちらも合わせて修正しました！
( items/index, items/show, orders/index関連を修正 )

iy06 reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

NUKEDO committed:
コードレビューを受けて修正

商品購入済みかどうかの記述のリファクタリング
Rubocop実行済

NUKEDO commented:
修正しました。

kokikurahashi commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,45 @@
+class OrdersController < ApplicationController
+  before_action :authenticate_user!, only: :index
+
+  def index
+    find_item
```
find_itemは別のアクションでも呼ばれているためbefore_actionで呼び出す形にしましょう！

NUKEDO replied:
先のレビューで言われていたのに、メソッドにまとめるだけで終わらせてしまってました。
ありがとうございます！修正しました。

kokikurahashi commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,45 @@
+
+  def index
+    find_item
+    if @item.purchase_record
+      redirect_to root_path
+    elsif current_user.id != @item.user_id
+      @order_form = OrderForm.new
+    else
+      redirect_to root_path
+    end
```
こちら@item.purchase_recordとcurrent_user.id == @item.user_idがどちらか一方trueの時に`redirect_to root_path`が実行されますので、orを用いて条件式をまとめましょう！

kokikurahashi commented on app/views/items/show.html.erb:
```diff
@@ -28,8 +27,8 @@
         <%= link_to '削除', "#", method: :delete, class:'item-destroy' %>
-      <% else %>
-        <%= link_to '購入画面に進む', "#" ,class:"item-red-btn"%>
+      <% elsif @stock %>
+        <%= link_to '購入画面に進む', item_orders_path(@item.id), method: :get, class:"item-red-btn"%>
```
@stock
こちら修正の名残だと思うので消しても問題ないことを確認しておきましょう！

NUKEDO replied:
修正忘れでした！ありがとうございます！
@stockは購入者ありの状態の時にtrue、購入者なしの状態でfalseを返す引数でした。
購入者情報なしの状態で購入画面に進むボタンが表示されるよう修正しました！

kokikurahashi commented on spec/models/order_form_spec.rb:
```diff
@@ -0,0 +1,137 @@
+      it 'トークンが存在し、全てのカラムに正しく入力されているので購入できる' do
+        expect(@order_form).to be_valid
+      end
+      it '郵便番号が正しく7桁分でありハイフン付きで入力されているので購入できる' do
+        @order_form.postal_code = '123-1234'
+        expect(@order_form).to be_valid
+      end
+      it '都道府県が2以上48以下なので購入できる' do
+        @order_form.prefecture_id = 48
+        expect(@order_form).to be_valid
+      end
+      it '建物名が空欄でも購入できる' do
+        @order_form.building_name = ''
+        expect(@order_form).to be_valid
+      end
+      it '電話番号が数字のみ10桁で入力されているので購入できる' do
+        @order_form.phone_number = '1234567890'
+        expect(@order_form).to be_valid
+      end
+      it '電話番号が数字のみ11桁で入力されているので購入できる' do
+        @order_form.phone_number = '12345678901'
+        expect(@order_form).to be_valid
+      end
+    end
```
建物名が空欄でも購入できる
以外は10~12行目でテスト済みのため不要です！

NUKEDO replied:
ありがとうございます！
25〜28行目についてはFactoryBotにて10桁で実行してますので削除しますが
29〜32行目の11桁については他で実行してませんので念のため入れておきます！
その他については修正しました！

kokikurahashi reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

NUKEDO committed:
コードレビューを受けて修正

NUKEDO commented:
プッシュをする前にコードレビュー依頼を出してしましました。
大変失礼いたしました。
修正しました！

kokikurahashi commented:
レビュー依頼ありがとうございます！
修正点問題なしです！
LGTMとさせていただきます！実装お疲れ様でした！
引き続き学習頑張っていきましょう！応援しております！
